base/src/main/java/com/calpano/graphinout/base/gio/ValidatingGioWriter.java

```diff
@@ -0,0 +1,103 @@
+package com.calpano.graphinout.base.gio;
+
+import lombok.RequiredArgsConstructor;
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.net.URL;
+import java.util.HashSet;
+import java.util.Set;
+
+@RequiredArgsConstructor
+public class ValidatingGioWriter implements GioWriter {
+    private final GioWriterImpl gioWriter;
+    private final Set<String> nodesIds = new HashSet<>();
+    private final Set<String> edgesIds = new HashSet<>();
+    private final Set<String> keysIds = new HashSet<>();
+    private final Set<String> endpointsNode = new HashSet<>();
+    private final Set<String> endpointsPort = new HashSet<>();
+    private final Set<String> nodePortName = new HashSet<>();
+
+    @Override
+    public void endDocument() throws IOException {
+        gioWriter.endDocument();
+    }
+
+    @Override
+    public void endEdge() throws IOException {
+        gioWriter.endEdge();
+    }
+
+    @Override
+    public void endGraph(@Nullable URL locator) throws IOException {
+        if (!nodesIds.containsAll(endpointsNode))
```

Comment:
Why are some checks in endGraph() and others in endDocument() -- should all be in endGraph(), I think